res/controllers/Behringer-Extension-scripts.js

```diff
@@ -4,10 +4,53 @@
 (function(global) {
 
     /** @private */
-    var components = global.components;
+    const components = global.components;
 
     /** @private */
-    var engine = global.engine;
+    const engine = global.engine;
+
+    /**
+     * Determines the merge strategy for a value when merging a component container definition.
+     *
+     * @param value anything
+     * @returns {boolean} merge strategy for the value: true for 'assign', false for 'deep merge'
+     * @private
+     */
+    const doAssign = value => {
+        return typeof value !== "object"
+            || value === null
+            || value instanceof components.ComponentContainer
+            || value instanceof components.Component;
+    };
+
+    /**
+     * Merges a list of component container definitions into a target object.
+     *
+     * This is not a generic deep merge algorithm for JS objects. It does not handle circular references.
+     * If a definition contains a reference to an object instance of a component or component container
+     * (e.g. `ShiftButton.target`), the instance is taken over by reference, no deep copy is made.
+     *
+     * @param {object} targetObject Target object for the merged definitions
+     * @param {Array} sources Source definitions
+     * @returns {object} The target object
+     * @see `GenericMidiController` on component container definition
+     * @private
+     */
+    const mergeDefinitions = (targetObject, ...sources) => sources.reduce((target, source) => {
```

Comment:
I don't understand… You need a deep merging algorithm after all?

Reply:
Yes, 1 of 4 uses of lodash's `_.merge()` requires a deep merge, the other 3 uses are working with a shallow merge. The scope is within `Behringer-Extension-scripts.js` only. It is used to merge mapping definitions of the user with defaults from the user and defaults from the mapping. Most mapping definitions simply contain data, but the `ShiftButton` may also reference a JS object (the target that is shifted by the button). This may cause a circular reference in the mapping description that must be considered in the merge algorithm.

This is the only case I found requiring a deep merge. No commonly used location (like `applyLayer()`) is affected anymore. Due to the limited scope, verifying the behavior was feasible.